subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AnnotationProcessorDetector.java

```diff
@@ -28,25 +28,22 @@
 import org.gradle.api.tasks.compile.CompileOptions;
 import org.gradle.cache.internal.FileContentCache;
 import org.gradle.cache.internal.FileContentCacheFactory;
-import org.gradle.internal.FileUtils;
-import org.gradle.internal.file.FileType;
 import org.gradle.internal.serialize.BaseSerializerFactory;
-import org.gradle.util.DeprecationLogger;
 
 import java.io.File;
-import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 
 public class AnnotationProcessorDetector {
     private final FileCollectionFactory fileCollectionFactory;
-    private final FileContentCache<Boolean> cache;
+    private final FileContentCache<Map<String, String>> cache;
```

Comment:
OK!  So after implementing your proposed change, I've finally remembered why I did it that way. :)

BaseSerializerFactory has a set of predefined types that can go into the FileContentCache.  The only reasonably rich data type it offers is NO_NULL_STRING_MAP_SERIALIZER, which made it really easy to represent the (currently three) properties I store per cache entry -- isProcessor, isIncremental, and processorName (for logging/blame).

The only other rich type it has is byte[], which could conceivably be used to store a serialized representation of AnnotationProcessorInfo by serializing it somehow (java serialization, maybe?)  But that's kind of a lot of work to store an opaque blob, particularly since it's fragile to changes to the AnnotationProcessorInfo class.  Whereas a Map<String, String> can be given reasonable default values for everything and is forward-compatible.

I think it might be best to keep the Map<String, String> as the cache value type, and just document it better.  Thoughts?

Reply:
You can provide serializers for any type. No need to restrict yourself to what the base serialiser provides.

Reply:
Done (for next push).